Inject Calypso's domain contact details form into composite checkout

nbloomf opened the pull request:
#### Changes proposed in this Pull Request

* Inject `ContactDetailsFormFields` as a managed component into composite checkout.

#### Testing instructions

Tests: `npm install && npm run test-packages composite-checkout-wpcom && npm run test-client client/my-sites/checkout/checkout`
Typecheck: `cd packages/composite-checkout-wpcom/ && tsc; cd ../..`

For manual testing:
1. Start calypso with composite checkout enabled using `ENABLE_FEATURES=composite-checkout-wpcom npm start`
2. Make sure you have the store and billing sandboxes enabled on the backend
3. Enter checkout with a domain in your cart.
4. Proceed to the contact details form. Fill out the form, checking that the `Continue` button does not become active until the form is complete. Watch the console for validation errors.
5. In the review step, delete the domain from your cart.
6. Return to the contact details step; verify that the fields there are already filled in.

Commit:
Add type signatures for DomainContactDetails and related structures

Commit:
Add type annotations to WPCOM store hook

Commit:
Add error boundary component

Commit:
Inject domain contact form

Commit:
Update composite-checkout in calypso

Commit:
(WIP) inject domain contact form validation

Commit:
Fix bug in call to renderDomainContactFields

sirbrillig commented on client/my-sites/checkout/checkout/composite-checkout.js:
```diff
@@ -124,6 +133,31 @@ export default function CompositeCheckout( {
+		applyDomainContactValidationResults
+	) => {
+		return (
+			<WPCheckoutErrorBoundary componentTitle="ContactDetailsFormFields">
```
What is the `componentTitle` prop for?

nbloomf replied:
Good catch- it is from an earlier iteration of WPCheckoutErrorBoundary that tried to be a little too clever, and is no longer needed.

sirbrillig commented on packages/composite-checkout-wpcom/src/components/wp-checkout-error-boundary.tsx:
```diff
@@ -0,0 +1,37 @@
+
+	render() {
+		if ( this.state.hasError ) {
+			return <h1>{ `Something went wrong in composite-checkout-wpcom.` }</h1>;
```
We might want to localize this message (and not mention the package name).

sirbrillig commented on packages/composite-checkout-wpcom/src/components/wp-contact-form.js:
```diff
@@ -14,11 +14,76 @@ import { useHasDomainsInCart } from '../hooks/has-domains';
 
-export default function WPContactForm( { summary, isComplete, isActive } ) {
-	const isDomainFieldsVisible = useHasDomainsInCart();
+const contactDetailsFormat = ( { hasDomainsInCart } ) => {
```
We could move this to the bottom of the file by changing it to use the `function` keyword instead of `const`. That way the main export of this file still comes first.

sirbrillig commented on packages/composite-checkout-wpcom/src/components/wp-contact-form.js:
```diff
@@ -14,11 +14,76 @@ import { useHasDomainsInCart } from '../hooks/has-domains';
+	return 'DEFAULT';
+};
+
+const renderContactDetails = ( {
```
Same with this function; it could be defined at the bottom of the file.

sirbrillig commented on packages/composite-checkout-wpcom/src/components/wp-contact-form.js:
```diff
@@ -14,11 +14,76 @@ import { useHasDomainsInCart } from '../hooks/has-domains';
+		default:
+			return (
+				<React.Fragment>
+					<TaxFields
```
We're still going to need the country drop-down, the phone number masking,  and the postal code validation for `TaxFields`. Is there some way we can leverage the domain contact fields to get these three?

nbloomf replied:
I see that you injected calypso's fields for these data in #38869, so I'm not sure how important this concern is now. But I think the short answer is no; I don't see a way to conditionally hide fields in the [ContactDetailsFormFields](https://github.com/Automattic/wp-calypso/tree/master/client/components/domains/contact-details-form-fields), and at the moment we do have less control over these fields for e.g. displaying error states.

Commit:
Fix bug in validation

Commit:
Fix type errors

Commit:
Fix broken tests

Commit:
Fix lint errors

Commit:
Incorporate feedback

Commit:
update package and package-lock

Commit:
Remove unused AddressFields component

Commit:
Suppress lint error

Commit:
Remove extraneous function

Commit:
Minor fixes

sirbrillig commented on packages/composite-checkout-wpcom/src/components/wp-checkout.js:
```diff
@@ -94,10 +96,11 @@ export default function WPCheckout( {
 			completeStepContent: <WPContactForm summary isComplete={ true } isActive={ false } />,
-			isCompleteCallback: () => isFormComplete( contactInfo, itemsWithTax ),
+			isCompleteCallback: () => isCompleteAndValid( contactInfo ),
 			isEditableCallback: () => isFormEditable( contactInfo, itemsWithTax ),
```
The second argument for `isFormEditable` was removed so you can remove it here (and does that mean that `itemsWithTax` is not used anywhere?)

sirbrillig commented on packages/composite-checkout-wpcom/src/components/wp-contact-form.js:
```diff
@@ -14,19 +14,22 @@ import { useHasDomainsInCart } from '../hooks/has-domains';
 	const isDomainFieldsVisible = useHasDomainsInCart();
 	const contactInfo = useSelect( select => select( 'wpcom' ).getContactInfo() );
 	const setters = useDispatch( 'wpcom' );
+	const domainName = 'example.com'; // TODO: get the actual domain name
```
```suggestion
	const [ items ] = useLineItems();
	const firstDomainItem = items.find( isLineItemADomain );
	const domainName = firstDomainItem ? firstDomainItem.sublabel : siteUrl;
```

For which you'll need to import `useLineItems` (from `@automattic/composite-checkout`) and `isLineItemADomain` (from `../hooks/has-domains`).

sirbrillig commented on packages/composite-checkout-wpcom/src/components/wp-contact-form.js:
```diff
@@ -89,11 +84,13 @@ const FieldRow = styled( GridRow )`
+	return countriesWithVAT.includes( country );
 }
 
+// TODO: Figure out if we still need this
```
No, you're right. This should be removed. The linters didn't find it because of its proptypes definition. It would make sense to remove it here. Thanks!

sirbrillig commented on packages/composite-checkout-wpcom/src/components/wp-contact-form.js:
```diff
@@ -174,135 +171,11 @@ const DomainRegistrationCheckbox = styled.input`
+	const { updatePhone, updatePhoneNumberCountry } = setters;
 
-	const isError = contactInfo.phoneNumber.isTouched && ! contactInfo.phoneNumber.isValid;
+	const isError = contactInfo.phone.isTouched && ! isValid( contactInfo.phone );
```
This is a reminder to myself to make sure that changing this to `phone` doesn't break the payment method submissions.

sirbrillig commented on packages/composite-checkout-wpcom/src/components/wp-contact-form.js:
```diff
@@ -359,7 +226,6 @@ const Label = styled.label`
 PhoneNumberField.propTypes = {
 	isRequired: PropTypes.bool,
 	id: PropTypes.string.isRequired,
-	setContactField: PropTypes.func.isRequired,
```
I believe you missed replacing this with `setters`:

```js
setters: PropTypes.shape({updatePhone: PropTypes.func.isRequired, updatePhoneNumberCountry: PropTypes.func.isRequired}).isRequired,
```

sirbrillig commented on packages/composite-checkout-wpcom/src/components/wp-contact-form.js:
```diff
@@ -531,8 +368,64 @@ function joinNonEmptyValues( joinString, ...values ) {
+						setters.updateContactDetails,
+						setters.applyDomainContactValidationResults
+					) }
+					{ requiresVatId && <VatIdField /> }
```
I'm surprised the VAT id isn't covered by the domain contact fields. 🤔

nbloomf replied:
I just checked again and it appears VAT is not included in the domain contact fields. I think that component is designed to collect the data needed by the registrar (it's used to let customers update their WHOIS record for instance) but VAT is only relevant for payments.

sirbrillig commented on packages/composite-checkout-wpcom/src/components/wp-contact-form.js:
```diff
@@ -400,11 +266,11 @@ function TaxFields( { section, taxInfo, setters, CountrySelectMenu, countriesLis
-					onChange={ value =>
-						setContactField( 'postalCode', { value, isTouched: true, isValid: !! value } )
-					}
+					onChange={ event => {
```
I think that `Field`'s `onChange` function already pulls out the `value` from the event object.

sirbrillig commented on packages/composite-checkout-wpcom/src/components/wp-contact-form.js:
```diff
@@ -531,8 +368,64 @@ function joinNonEmptyValues( joinString, ...values ) {
-const StateSelectWrapper = styled.div`
-	& select {
-		width: 100%;
+function contactDetailsFormat( { isDomainFieldsVisible } ) {
```
Minor, but might this be more readable with a verb like `getContactDetailsFormat`? As it is I first thought this was supposed to format the contact details.

sirbrillig commented on packages/composite-checkout-wpcom/src/components/wp-contact-form.js:
```diff
@@ -531,8 +368,64 @@ function joinNonEmptyValues( joinString, ...values ) {
+	CountrySelectMenu,
+	countriesList,
+} ) {
+	const format = contactDetailsFormat( { isDomainFieldsVisible } );
```
If there's only one argument, we probably can pass it without an object wrapper and destructuring.

sirbrillig commented on packages/composite-checkout-wpcom/src/types/wpcom-store-state.ts:
```diff
@@ -0,0 +1,271 @@
+	organization: ManagedValue< string >;
+	email: ManagedValue< string >;
+	alternateEmail: ManagedValue< string >;
+	phone: ManagedValue< string >;
```
Since the names of these fields have changed (eg: `phoneNumber` to `phone`, `country` to `countryCode`), they will break anywhere that currently uses `getContactInfo()`, notably the payment method getters defined in `createPaymentMethods()`. We should find those places and change them too.

nbloomf replied:
I think I fixed all of these.

sirbrillig reviewed (commented):
Great work. I've left some comments, but they're mostly minor things.

The main thing that bothers me is that the fields are only validated on blur rather than while typing. I wonder if we can fix that here, but it could be a future PR.

nbloomf commented on client/my-sites/checkout/checkout/composite-checkout.js:
```diff
@@ -180,6 +220,7 @@ export default function CompositeCheckout( {
 					countriesList={ countriesList }
 					PhoneInput={ PhoneInput }
 					StateSelect={ StateSelect }
+                    renderDomainContactFields={ renderDomainContactFields }
```
I've had issues like this before -- that line was edited during an interactive rebase, so prettier probably didn't even run before the commit.

nbloomf commented:
> The main thing that bothers me is that the fields are only validated on blur rather than while typing. I wonder if we can fix that here, but it could be a future PR.

This does make it awkward when you finish typing into the last field and the continue button does not become active until the field is blurred. OTOH validation here is done on the server, so (naively) revalidating on every edit would cause a lot of extra http requests. It would be interesting if the validation could be done on every edit, but debounced to prevent too many unnecessary requests.

nbloomf commented on client/my-sites/checkout/checkout/test/composite-checkout.js:
```diff
@@ -91,6 +98,7 @@ describe( 'CompositeCheckout', () => {
 					getCart={ mockGetCartEndpointWith( initialCart ) }
 					getStoredCards={ async () => [] }
 					allowedPaymentMethods={ [ 'paypal' ] }
+					contactCountriesList={ countryList }
```
You're right, that is left over from an earlier refactoring. It's a little disturbing that this test didn't fail.

nbloomf commented on packages/composite-checkout-wpcom/src/components/wp-checkout-error-boundary.tsx:
```diff
@@ -0,0 +1,39 @@
+ */
+import React from 'react';
+
+type WPCheckoutErrorBoundaryState = {
```
I agree that this is a good idea. However unless you have a strong objection I think I'll leave this where it is for now; we need to get this merged and have talked about the need to move other components into calypso for the same reason.

sirbrillig commented:
Yeah, I didn't know it did the validation server-side. That makes sense. We can push that off to a later PR.

Commit:
Remove unnecessary change to package.json

Commit:
Fix whitespace

Commit:
Add prepareDomainContactDetails to package exports for composite-checkout-wpcom

Commit:
Update contact details field names in composite-checkout-payment-methods

Commit:
Update prop name in test

Commit:
Remove unused function parameter

Commit:
Get domain name from cart in contact form

Commit:
Remove unused components in contact form

Commit:
Update proptypes

Commit:
Pass siteUrl into WPContactForm

Commit:
Use correct parameter type in onChange handler

Commit:
Add domain dependencies to WPContactForm

Commit:
Change function signature to plain arguments rather than destructuring

Commit:
Pass countriesList into ContactDetailsFormFields

sirbrillig reviewed (commented):
It looks like it's mostly working well! I was examining the network requests when sending a transaction and it looks as though the country code, while being correctly provided as part of `domain_details`, is not being included at all in `payment` or `cart.tax.location`. That means that it's not set when being passed to `formatDataForTransactionsEndpoint()`.

The debug logs (I used an existing card payment method) show that the data submitted to `submitExistingCardPayment()` included `country`, but not `countryCode`. Indeed, I see that `beginStripeTransaction()` in `createExistingCardMethod`, `createStripeMethod`, and the rest sets the key to `country`. I think it needs to be changed there too. Likely `phoneNumber` as well.

To be honest, there's too many combinations of `country` and `countryCode` for me to keep track of. It might be worth changing them all to one or the other except where absolutely required. I believe the transactions endpoints expect `countryCode` or `country_code` for all the fields where the country is sent, and the paypal endpoint expects `country`.

Commit:
Move composite checkout type exports to an index file

Commit:
Collapse type exports

nbloomf commented:
> To be honest, there's too many combinations of country and countryCode for me to keep track of.

Agreed. :) In #39007 we added typescript annotations to these parameter objects, and I just rebased on top of those.

Commit:
Fix duplicated import

sirbrillig commented on client/my-sites/checkout/checkout/composite-checkout.js:
```diff
@@ -155,6 +164,38 @@ export default function CompositeCheckout( {
+					onValidate={ ( values, onComplete ) => {
+						// TODO: Should probably handle HTTP errors here
+						validateDomainContact( values, domainNames, ( httpErrors, data ) => {
+							if ( data.messages ) {
```
This is fine but five lines for a debug message seems like a lot. Maybe something like this would be a little more compact:

```js
debug( 'Domain contact info validation ' + data.messages ? 'errors' : 'successful', data.messages );
```

nbloomf replied:
Prettier expands this to four lines :/

(Albeit with only one expression, which is still an improvement)

sirbrillig commented on client/my-sites/checkout/checkout/composite-checkout-payment-methods.tsx:
```diff
@@ -111,10 +112,10 @@ export function createPaymentMethods( {
 							domainDetails: getDomainDetails( select ),
 							couponId: null, // TODO: get couponId
-							country: select( 'wpcom' )?.getContactInfo?.()?.country?.value,
+							countryCode: select( 'wpcom' )?.getContactInfo?.()?.countryCode?.value,
```
This is submitting `countryCode` to `makePayPalExpressRequest()`, but that function is looking for `country`, so it looks like this is still an issue.

sirbrillig reviewed (commented):
Aside: I'm getting some typescript errors in my editor in `client/my-sites/checkout/checkout/types/paypal-express.ts` and `client/my-sites/checkout/checkout/types/transaction-endpoint.ts`:

```
debug: ( string, any ) => void; # [tsserver] Parameter has a name but no type. Did you mean 'arg0: string'? [7051]
```

```
export type WPCOMTransactionEndpoint = (
	WPCOMTransactionEndpointRequestPayload # [tsserver] Parameter has a name but no type. Did you mean 'arg0: WPCOMTransactionEndpointRequestPayload'? [7051]
) => Promise< WPCOMTransactionEndpointResponse >;
```

```
	const currency = items.reduce( ( firstValue, item ) => firstValue || item.amount.currency, null ); # [tsserver] No overload matches this call.  Overload 1 of 3, '(callbackfn: (previousValue: WPCOMCartItem, currentValue: WPCOMCartItem, currentIndex: number, array: WPCOMCartItem[]) => WPCOMCartItem, initialValue: WPCOMCartItem): WPCOMCar..
.f 3, '(callbackfn: (previousValue: null, currentValue: WPCOMCartItem, currentIndex: number, array: WPCOMCartItem[]) => null, initialValue: null): null', gave the following error.    Type 'string' is not assignable to type 'null'.
```

There's others as well. Are those known issues? Or is my editor being weird?

Commit:
Make debug statement more compact

Commit:
Fix parameter name in paypal request payload

Commit:
Fix type warnings

nbloomf commented:
> Aside: I'm getting some typescript errors in my editor... are these known issues?

Good catch. Because this file is in calypso, type errors are not fatal -- I should have realized this. :) Additionally the typescript compiler runs out of memory on my machine. Fortunately the full list of errors is available in the CI logs.

Some of these errors are not great but can be ignored for now because `composite-checkout-payment-methods` started its life as untyped JS -- particularly the `implicit any` warnings. Others, including the three you quoted above, can and should be fixed.

Commit:
Remove PhoneNumberField from plan-only contact form

sirbrillig committed:
Use `country` when calling submitCreditsTransaction

sirbrillig committed:
Remove phone number from everywhere but domain details

sirbrillig commented:
I noticed a few things which I pushed some changes for... but when I went to test it just now I'm getting a weird validation error and I haven't even tried adding a domain yet

![Screen Shot 2020-01-24 at 5 07 28 PM](https://user-images.githubusercontent.com/2036909/73107569-21d3ca80-3ecc-11ea-8675-3a9e0384e5aa.png)

Let me see what's going on.

sirbrillig commented:
I figured it out; by default each managed value had an error included (an empty string). I assume that was there to make it so that untouched items are not valid for the step's check (`isCompleteAndValid`), but when editing the non-domain fields there's no validator (I think) to set the errors to be empty so I tried removing it. However, then that made it so that the form was valid even if nothing was entered. I need to think this through more carefully.

sirbrillig committed:
Use isValid for vatId

sirbrillig committed:
Add propTypes to TaxFields setters

sirbrillig commented:
I keep mixing it up in my head, but let me write down the things we need from validation so I can keep it straight:

- If a field is not touched, it must not display errors, even if it has some.
- If a field is touched, it should display errors if it has some.
- If a field has errors, it is invalid and the continue button should be disabled.
- If a field is empty, it is invalid and the continue button should be disabled.

For the latter two, we have `isCompleteAndValid` which is called by the step's `isCompleteCallback` property.

sirbrillig committed:
Validate postalCode and countryCode when setting

sirbrillig committed:
Make ManagedValue always a string

sirbrillig committed:
initialize ManagedValues with objects

This way their errors value can be easily overridden if the field is
optional.

sirbrillig commented:
This _should_ work now, I think. I added validation to `TaxFields` fields. However, now there's a TypeScript issue where it seems to think we're only using es5; despite being able to use many newer features like object spreads and optional chaining, we cannot use `Object.values`. I'm researching the answer.

sirbrillig committed:
Explicitly add ES2017 to TS lib

sirbrillig commented:
[This comment](https://github.com/microsoft/TypeScript/issues/28166#issuecomment-433564613) helped me make sense of the TS error. We're using `lib` so we have to specify the libs we need, no matter what the `target` is set to.

I think this resolves all the issues; I'm correctly getting validation errors when I should in the non-domains version of the form. The domains version of the form continues to frustrate me with how long it takes to validate, but that's expected (notably, if you fill in a field correctly, then empty the field, you can press "Continue" before the validation finishes; we should disable the "Continue" button while validation is running or make pressing it an async operation). 

Now I noticed that the domains form does not ever seem to display "This field is required" errors. I hope I didn't break that somehow.

I also removed some generics because they seemed like a premature optimization, but maybe we need them for something? The only values I could find were strings; are there others?

sirbrillig commented:
To be clear, the main reason I removed the generic is so we have access to `value.length`. 

As for the Continue button, don’t worry about it now. I have some ideas about making it async and more flexible which should resolve this. 

Therefore I think the only thing still broken here is displaying validation errors in the domain form.

nbloomf commented:
> I also removed some generics because they seemed like a premature optimization, but maybe we need them for something? The only values I could find were strings; are there others?

If `ManagedValue` lived in a separate library I would keep it polymorphic. The only managed values here are strings, so specializing to `string` shouldn't cause any problems.

Commit:
Fix associativity of ternary if

nbloomf commented on packages/composite-checkout-wpcom/src/types/wpcom-store-state.ts:
```diff
@@ -15,15 +15,20 @@ export function isValid( arg: ManagedValue ): boolean {
 }
 
-function initialManagedValue( value: string ): ManagedValue {
+function getInitialManagedValue( initialProperties?: {
```
This is probably bikeshedding, but I'm of two minds on using the `get` prefix for a pure function like this. It probably comes from working on the backend, but I instinctively expect the act of getting to touch a database at some point. :)

sirbrillig replied:
This is totally a subjective discussion so I'll leave it to you to decide. 

My opinion is that functions should always start with a verb firstly to make it clear that the object is a function and not just a variable (I might write `const myValue = initialManagedValue` accidentally and it looks right - much less relevant in a language like Ruby where functions can be called without parenthesis) and also to help [clarify ambiguity around the function's purpose](https://github.com/Automattic/phpcs-neutron-standard/blob/master/GUIDELINES.md#function-naming) (although in this case it's pretty clear what the function does).

sirbrillig commented:
Any thoughts about how to get validation errors to display inline? Is this something we should postpone for another PR?

nbloomf commented:
> Any thoughts about how to get validation errors to display inline?

This should be straightforward, but as we discussed on Slack it may be better to do that in a separate PR. This one is already enormous.

sirbrillig reviewed (approved):
:shipit: We still have more to do here but this is a good improvement.
